pkg/virt-controller/watch/instancetype/upgrade.go

```diff
@@ -0,0 +1,241 @@
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+package instancetype
+
+import (
+	"fmt"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+
+	instancetypev1beta1 "kubevirt.io/api/instancetype/v1beta1"
+
+	"kubevirt.io/kubevirt/pkg/controller"
+	"kubevirt.io/kubevirt/pkg/instancetype"
+	"kubevirt.io/kubevirt/pkg/util/status"
+)
+
+type UpgradeController struct {
+	Queue             workqueue.RateLimitingInterface
+	upgrader          instancetype.UpgraderInterface
+	statusUpdater     *status.ControllerRevisionUpgradeStatusUpdater
+	recorder          record.EventRecorder
+	crInformer        cache.SharedIndexInformer
+	crUpgradeInformer cache.SharedIndexInformer
+}
+
+func NewUpgradeController(
+	client kubecli.KubevirtClient,
+	recorder record.EventRecorder,
+	vmInformer cache.SharedIndexInformer,
+	crInformer cache.SharedIndexInformer,
+	crUpgradeInformer cache.SharedIndexInformer,
+) (*UpgradeController, error) {
+	c := &UpgradeController{
+		Queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "virt-controller-instancetype-migration"),
+		upgrader:          instancetype.NewUpgrader(client, vmInformer),
+		statusUpdater:     status.NewControllerRevisionUpgradeStatusUpdater(client),
+		recorder:          recorder,
+		crInformer:        crInformer,
+		crUpgradeInformer: crUpgradeInformer,
+	}
+
+	_, err := c.crUpgradeInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addUpgrade,
+			UpdateFunc: c.updateUpgrade,
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *UpgradeController) enqueueUpgrade(obj interface{}) {
+	crUpgrade := obj.(*instancetypev1beta1.ControllerRevisionUpgrade)
+	key, err := controller.KeyFunc(crUpgrade)
+	if err != nil {
+		logger := log.Log
+		logger.Object(crUpgrade).Reason(err).Error("Failed to extract key from ControllerRevisionUpgrade.")
+		return
+	}
+	c.Queue.Add(key)
+}
+
+func (c *UpgradeController) addUpgrade(obj interface{}) {
+	c.enqueueUpgrade(obj)
+}
+
+func (c *UpgradeController) updateUpgrade(_, currObj interface{}) {
+	c.enqueueUpgrade(currObj)
+}
+
+func (c *UpgradeController) Run(threadiness int, stopCh <-chan struct{}) {
+	defer controller.HandlePanic()
+	defer c.Queue.ShutDown()
+
+	log.Log.Info("Starting ControllerRevisionUpgrade controller.")
+	defer log.Log.Info("Stopping ControllerRevisionUpgrade controller.")
+
+	cache.WaitForCacheSync(stopCh, c.crInformer.HasSynced, c.crUpgradeInformer.HasSynced)
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+
+}
+
+func (c *UpgradeController) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *UpgradeController) Execute() bool {
+	key, quit := c.Queue.Get()
+	if quit {
+		return false
+	}
+	defer c.Queue.Done(key)
```

Comment:
Did I understand correctly that Queue can contain a key multiple times? So when it is reenqueued this does not remove the reenqueued key?

Reply:
No I think it's a single key but `AddRateLimited` / `Add` [1] marks the key as dirty (requires processing) that is the handled by `Done` [2]  with the key added back into the queue again. ewwww.

[1]  https://github.com/kubevirt/kubevirt/blob/1b9232fada18fb9dfaac67a73a2ea880f188e5e3/vendor/k8s.io/client-go/util/workqueue/queue.go#L119-L139
[2] https://github.com/kubevirt/kubevirt/blob/1b9232fada18fb9dfaac67a73a2ea880f188e5e3/vendor/k8s.io/client-go/util/workqueue/queue.go#L177-L193